casper/src/main/resources/AsyncMap.rho

```diff
@@ -0,0 +1,208 @@
+// Rholang map desiderata: speedy insert & lookup,  no conflicts on lookup, no conflicts on inserts to different keys
+// This implementation: O(log n) insert & lookup; also provides O(1) lookup when it is known that the value at a key exists.
+// Conflict analysis
+//   Lookup
+//     When looking up a value, only peeks are used, so lookups will not conflict.
+//   Insert
+//     When inserting, only peeks are used on existing cells (except the last
+//     shared one in the path), while newly created cells have a different name.
+//     So there's conflict only if the keys share a common prefix that hadn't
+//     already been populated.
+// Usage
+// new o(`rho:io:stdout`), mapCh in {
+//   o!("Initializing map") |
+//   // Use 3 * 8 = 24 bits of parallelization
+//   AsyncMap!("init", 3, *mapCh) |
+//   for (@map <- mapCh) {
+//     o!("Map initialized, setting") |
+//     new ret1, ret2, ret3 in {
+//       AsyncMap!("set", map, "some key", "some val", *ret1) |
+//       AsyncMap!("set", map, "monkey", "some other val", *ret2) |
+//       for (_ <- ret1; _ <- ret2) {
+//         o!("Value set, getting") |
+//         AsyncMap!("get", map, "some key", *ret1) |
+//         AsyncMap!("fastUnsafeGet", map, "monkey", *ret2) |
+//         AsyncMap!("get", map, "some unused key", *ret3) |
+//         for (@val1 <- ret1; @val2 <- ret2; @val3 <- ret3) {
+//           // Should print ["Got these from the map: ", ["some val"], ["some other val"], []]
+//           o!(["Got these from the map: ", val1, val2, val3])
+//         }
+//       }
+//     }
+//   }
+// }
+
+new AsyncMap, o(`rho:io:stdout`) in {
+  new MakeCell, ByteArrayToNybbleList, AsyncMapSetter, AsyncMapGetter, powersCh in {
+    powersCh!([1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768,65536]) |
+    for (@powers <- powersCh) {
+      contract MakeCell(@initVal, cell) = {
+        new cellStore in {
+          cellStore!(initVal) |
+          contract cell(@"get", ret) = {
```

Comment:
We should probably parameterize the cell-methods with a 'this' pointer and only create them once to avoid TS bloat.

Also, can we name it a Node? I think that's what it is :) So we'd have:

```
contract MakeNode(@val, prefix, store)
contract nodeGet(store, ret)
contract nodeSetBit(store, @bit, ack)
...
```

Reply:
Done